### PR TITLE
case weights

### LINKED ISSUE #136: No support for weights
The `weights` argument in `fit` does nothing. Most modeling packages allow for case weights, so is this going to be supported in the future?

### PR DESCRIPTION
**Pull request by topepo**
Closes #136

### CODE REVIEW

**Comment by topepo** [comment]
All models with case weights will be tested in extratests. 

Spark support is still outstanding because there are some API issues to work out.  Sparks models want a column name for the weights argument.

File: R/logistic_reg_data.R [hunk 1]
```diff
@@ -220,7 +220,7 @@ set_fit(
   mode = "classification",
   value = list(
     interface = "matrix",
-    protect = c("x", "y", "wi"),
+    protect = c("x", "y"),
```
Comment (DavisVaughan) [inline R/logistic_reg_data.R:223]: What was going on here?

File: R/misc.R [hunk 2]
```diff
@@ -385,3 +385,18 @@ stan_conf_int <- function(object, newdata) {
 
   penalty
 }
+
+
+check_case_weights <- function(x, spec) {
+  if (is.null(x)) {
+    return(invisible(NULL))
+  }
+  if (!inherits(x, "hardhat_case_weights")) {
```
Comment (DavisVaughan) [inline R/misc.R:394]:
```suggestion
  if (!hardhat::is_case_weights(x)) {
```

File: tests/testthat/test-case-weights.R [hunk 3]
```diff
@@ -0,0 +1,57 @@
+
+test_that('case weights with xy method', {
+
+  skip_if_not_installed("C50")
+  skip_if_not_installed("modeldata")
+  library(hardhat)
```
Comment (DavisVaughan) [inline tests/testthat/test-case-weights.R:6]: Probably best not to `library(hardhat)`

File: tests/testthat/test-case-weights.R [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+  data("two_class_dat", package = "modeldata")
+
+  wts <- runif(nrow(two_class_dat))
+  wts <- ifelse(wts < 1/5, 0, 1)
+  two_class_subset <- two_class_dat[wts != 0, ]
+  wts <- importance_weights(wts)
```
Comment (DavisVaughan) [inline tests/testthat/test-case-weights.R:12]: parsnip is one of the packages where I think it makes sense to re-export both `frequency_weights()` and `importance_weights()`. That way users get access to them and you can use them in tests without needing to library hardhat or prefix them with `hardhat::`

File: tests/testthat/test-case-weights.R [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+      set_engine("C5.0") %>%
+      set_mode("classification") %>%
+      fit(Class ~ ., data = two_class_dat, case_weights = wts)
+  },
+  regexp = NA)
+
+  expect_output(
+    print(C5_bst_wt_fit$fit$call),
+    "weights = weights"
+  )
```
Comment (DavisVaughan) [inline tests/testthat/test-case-weights.R:27]: Is there anything inside the C5 model object that you can check for to prove that you are using case weights? That might be more robust

**Review by DavisVaughan — COMMENTED** [review]
Looking pretty good so far!

File: tests/testthat/test-case-weights.R (hunk 3 above)
Comment (topepo) [reply tests/testthat/test-case-weights.R:6]: Should the individual weight functions be reexported?

File: tests/testthat/test-case-weights.R (hunk 4 above)
Comment (topepo) [reply tests/testthat/test-case-weights.R:12]: OK this answers ☝️

File: tests/testthat/test-case-weights.R (hunk 5 above)
Comment (topepo) [reply tests/testthat/test-case-weights.R:27]: No, not really. There are a lot of models that don't keep records of the weights.  Also, with trees and earth models, the subsetting weights that I use don't produce the same results as the model when those rows are specifically absent. 

My thinking is, for those cases, I'll make sure that the weights were used in the call and the model didn't fail.

File: R/arguments.R [hunk 6]
```diff
@@ -165,6 +169,13 @@ make_form_call <- function(object, env = NULL) {
   # sub in actual formula
   fit_args[[ unname(data_args["formula"]) ]]  <- env$formula
 
+  # Add in case weights symbol
+  if (uses_weights) {
+    fit_args[[ unname(data_args["weights"]) ]]  <- rlang::expr(weights)
+  }
```
[the remark this answers is not in the document]
Comment (topepo) [reply R/arguments.R]: I added a lot of comments (that should feed back into tidymodels.org) that explain what these elements do. I cleaned up the code a little too in the process.

**Comment by DavisVaughan** [comment]
- [ ] Attempt to remove `convert_case_weights()` because it doesn't look to be actually used anymore

**Pull request merged**